

Here we are Experiment on all the hyperparameter tuning techniques on the    binary classification diabetes dataset for Neural Network model using Keras & TensorFlow libraries and finally and we can able to see the all the scores.

- Batch Size

  The number of samples trained before the model was updating.

- Epoch

  On epoch trains the entire dataset, In the training process it updates the weights based on the gradient.

- Learning Rate

 The weights of the learning rate network is updated during training and it controls the step size and gives better convergence.

- Dropout Rate

  dropout rate means it will drop the units randomly from neural network to prevent overfitting.

- Activation Function

  Activation function are applied output to each neuron to check whether it should be active or not? some common activation function are sigmoid, tanh, relu.

- Kernel Initializer

  It initializes the weights to neural network with proper process it will speed up the convergence & improve performance.

- Neurons

  Neurons means can changes the capacity & complexity of model.

- Activation Layer

  In Neural Network the activation function are applied to activation layers which introduce the non-linearity and learns the complex patterns.


# **Finally, Tuning these hyperparameters on model we can get the better performance.**













In [1]:
# Importing the Necessary libraries

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import keras
from sklearn.preprocessing import StandardScaler
import warnings
warnings.filterwarnings("ignore")

In [2]:
import pandas as pd

filename = "pima-indians-diabetes.data.csv"
names = ["preg", "plas", "pres", "skin", "test", "mass", "pedi", "age", "class"]
df = pd.read_csv(filename, names=names)
array = df.values


In [3]:
df.head()

preg  plas  pres  skin  test  mass   pedi  age  class
0     6   148    72    35     0  33.6  0.627   50      1
1     1    85    66    29     0  26.6  0.351   31      0
2     8   183    64     0     0  23.3  0.672   32      1
3     1    89    66    23    94  28.1  0.167   21      0
4     0   137    40    35   168  43.1  2.288   33      1

In [4]:
# Splitting the Data into Training and Testing
X=array[:,0:8]
y=array[:,8]

In [5]:
X

array([[  6.   , 148.   ,  72.   , ...,  33.6  ,   0.627,  50.   ],
       [  1.   ,  85.   ,  66.   , ...,  26.6  ,   0.351,  31.   ],
       [  8.   , 183.   ,  64.   , ...,  23.3  ,   0.672,  32.   ],
       ...,
       [  5.   , 121.   ,  72.   , ...,  26.2  ,   0.245,  30.   ],
       [  1.   , 126.   ,  60.   , ...,  30.1  ,   0.349,  47.   ],
       [  1.   ,  93.   ,  70.   , ...,  30.4  ,   0.315,  23.   ]])

In [6]:
y

array([1., 0., 1., 0., 1., 0., 1., 0., 1., 1., 0., 1., 0., 1., 1., 1., 1.,
       1., 0., 1., 0., 0., 1., 1., 1., 1., 1., 0., 0., 0., 0., 1., 0., 0.,
       0., 0., 0., 1., 1., 1., 0., 0., 0., 1., 0., 1., 0., 0., 1., 0., 0.,
       0., 0., 1., 0., 0., 1., 0., 0., 0., 0., 1., 0., 0., 1., 0., 1., 0.,
       0., 0., 1., 0., 1., 0., 0., 0., 0., 0., 1., 0., 0., 0., 0., 0., 1.,
       0., 0., 0., 1., 0., 0., 0., 0., 1., 0., 0., 0., 0., 0., 1., 1., 0.,
       0., 0., 0., 0., 0., 0., 0., 1., 1., 1., 0., 0., 1., 1., 1., 0., 0.,
       0., 1., 0., 0., 0., 1., 1., 0., 0., 1., 1., 1., 1., 1., 0., 0., 0.,
       0., 0., 0., 0., 0., 0., 0., 1., 0., 0., 0., 0., 0., 0., 0., 0., 1.,
       0., 1., 1., 0., 0., 0., 1., 0., 0., 0., 0., 1., 1., 0., 0., 0., 0.,
       1., 1., 0., 0., 0., 1., 0., 1., 0., 1., 0., 0., 0., 0., 0., 1., 1.,
       1., 1., 1., 0., 0., 1., 1., 0., 1., 0., 1., 1., 1., 0., 0., 0., 0.,
       0., 0., 1., 1., 0., 1., 0., 0., 0., 1., 1., 1., 1., 0., 1., 1., 1.,
       1., 0., 0., 0., 0.

In [7]:
# Standardization the data
S=StandardScaler()
S.fit(X)
X_standardized=S.transform(X)

In [8]:
pd.DataFrame(X_standardized).describe()

0             1             2             3             4  \
count  7.680000e+02  7.680000e+02  7.680000e+02  7.680000e+02  7.680000e+02   
mean  -6.476301e-17 -9.251859e-18  1.503427e-17  1.006140e-16 -3.006854e-17   
std    1.000652e+00  1.000652e+00  1.000652e+00  1.000652e+00  1.000652e+00   
min   -1.141852e+00 -3.783654e+00 -3.572597e+00 -1.288212e+00 -6.928906e-01   
25%   -8.448851e-01 -6.852363e-01 -3.673367e-01 -1.288212e+00 -6.928906e-01   
50%   -2.509521e-01 -1.218877e-01  1.496408e-01  1.545332e-01 -4.280622e-01   
75%    6.399473e-01  6.057709e-01  5.632228e-01  7.190857e-01  4.120079e-01   
max    3.906578e+00  2.444478e+00  2.734528e+00  4.921866e+00  6.652839e+00   

                  5             6             7  
count  7.680000e+02  7.680000e+02  7.680000e+02  
mean   2.590520e-16  2.451743e-16  1.931325e-16  
std    1.000652e+00  1.000652e+00  1.000652e+00  
min   -4.060474e+00 -1.189553e+00 -1.041549e+00  
25%   -5.955785e-01 -6.889685e-01 -7.862862e-01  
50%    9.419788e-04 -3.001282e-01 -3.608474e-01  
75%    5.847705e-01  4.662269e-01  6.602056e-01  
max    4.455807e+00  5.883565e+00  4.063716e+00

# **Hyperparameter Tuning for Batch Size & Epochs**

In [9]:
!pip install keras==2.12.0

In [10]:
from sklearn.model_selection import GridSearchCV   # It is used for to select best hyperparamenters values
from keras.models import Sequential                # It is used to build the Neural Network Model
from keras.layers import Dense                     # Hidden layers
from keras.wrappers.scikit_learn import KerasClassifier  # We have to put this Neural Network model into Keras classifier and we have to run the model
from sklearn.model_selection import KFold, GridSearchCV
from tensorflow.keras.optimizers import Adam   # Adam Optimizer

In [11]:
# Function to create the neural network model
def create_model():
    model = Sequential()
    model.add(Dense(12, input_dim=8, kernel_initializer='random_uniform', activation='relu'))
    model.add(Dense(8, kernel_initializer='random_uniform', activation='relu'))
    model.add(Dense(1, kernel_initializer='random_uniform', activation='sigmoid'))

    # Compiling the model
    adam = Adam(learning_rate=0.01)
    model.compile(loss='binary_crossentropy', optimizer=adam, metrics=['accuracy'])
    return model

In [12]:
# Define the Keras classifier
model = KerasClassifier(build_fn=create_model, verbose=0)

# Define the grid search parameters
batch_size = [10, 20, 40]
epochs = [10, 50, 100]

# Making the dictonary for Grid Search Parameters
param_grid = dict(batch_size=batch_size, epochs=epochs)

# Build and fit the grid search CV
grid = GridSearchCV(estimator=model, param_grid=param_grid, cv=KFold(), verbose=10)
grid_result = grid.fit(X_standardized, y)

Fitting 5 folds for each of 9 candidates, totalling 45 fits
[CV 1/5; 1/9] START batch_size=10, epochs=10....................................
[CV 1/5; 1/9] END .....batch_size=10, epochs=10;, score=0.740 total time=   3.6s
[CV 2/5; 1/9] START batch_size=10, epochs=10....................................
[CV 2/5; 1/9] END .....batch_size=10, epochs=10;, score=0.695 total time=   2.4s
[CV 3/5; 1/9] START batch_size=10, epochs=10....................................
[CV 3/5; 1/9] END .....batch_size=10, epochs=10;, score=0.773 total time=   2.5s
[CV 4/5; 1/9] START batch_size=10, epochs=10....................................
[CV 4/5; 1/9] END .....batch_size=10, epochs=10;, score=0.843 total time=   2.4s
[CV 5/5; 1/9] START batch_size=10, epochs=10....................................
[CV 5/5; 1/9] END .....batch_size=10, epochs=10;, score=0.758 total time=   5.7s
[CV 1/5; 2/9] START batch_size=10, epochs=50....................................
[CV 1/5; 2/9] END .....batch_size=10, epochs=50;,

In [13]:
# Summarize the results
print("Best:{},using{}".format(grid_result.best_score_,grid_result.best_params_))
means=grid_result.cv_results_["mean_test_score"]
stds=grid_result.cv_results_["std_test_score"]
params=grid_result.cv_results_["params"]
for mean,stdev,param in zip(means,stds,params):
  print("{},{} with:{}".format(mean,stdev,param))

Best:0.7735166907310486,using{'batch_size': 40, 'epochs': 50}
0.7618198871612549,0.048374817610974695 with:{'batch_size': 10, 'epochs': 10}
0.750029718875885,0.033588001534738884 with:{'batch_size': 10, 'epochs': 50}
0.7423054099082946,0.05174821953431119 with:{'batch_size': 10, 'epochs': 100}
0.7683218836784362,0.04662664853939113 with:{'batch_size': 20, 'epochs': 10}
0.7527204871177673,0.04688042853174413 with:{'batch_size': 20, 'epochs': 50}
0.7605211973190308,0.037902864913757564 with:{'batch_size': 20, 'epochs': 100}
0.7618113994598389,0.04164509127633645 with:{'batch_size': 40, 'epochs': 10}
0.7735166907310486,0.03006518653458518 with:{'batch_size': 40, 'epochs': 50}
0.7605211853981018,0.04048484277431093 with:{'batch_size': 40, 'epochs': 100}


- From the above results we can able to see that from what Batch Size -20 & Which Epoch - 10 is giving the Highest accuracy " Best:0.7657329797744751 "

# **Hyperparameter Tuning for Learning Rate & Dropout Rate**

- Dropout - We can drop the neurons from the hidden layers which can prevent the model from overfitting.

In [14]:
from keras.layers import Dropout

# Function to create and define the neural network model
def create_model(learning_rate,dropout_rate):
    model = Sequential()
    model.add(Dense(8, input_dim=8, kernel_initializer="normal", activation="relu"))
    model.add(Dropout(dropout_rate))
    model.add(Dense(4,input_dim=8,kernel_initializer="normal", activation="relu"))
    model.add(Dropout(dropout_rate))
    model.add(Dense(1, activation='sigmoid'))

    # Compiling the model
    adam = Adam(learning_rate=learning_rate)
    model.compile(loss='binary_crossentropy', optimizer=adam, metrics=['accuracy'])
    return model

# Create the model
model = KerasClassifier(build_fn=create_model, verbose=0, batch_size=40, epochs=10)

# Define the grid search parameters
learning_rate= [0.001,0.01,0.1]
dropout_rate = [0.0,0.1,0.2]

# Make a dictionary of grid search parameters
param_grids = dict(learning_rate=learning_rate,dropout_rate=dropout_rate)

# Build and fit the grid search CV
grid = GridSearchCV(estimator=model, param_grid=param_grids, cv=KFold(), verbose=10)
grid_result = grid.fit(X_standardized, y)


Fitting 5 folds for each of 9 candidates, totalling 45 fits
[CV 1/5; 1/9] START dropout_rate=0.0, learning_rate=0.001.......................
[CV 1/5; 1/9] END dropout_rate=0.0, learning_rate=0.001;, score=0.760 total time=   2.0s
[CV 2/5; 1/9] START dropout_rate=0.0, learning_rate=0.001.......................
[CV 2/5; 1/9] END dropout_rate=0.0, learning_rate=0.001;, score=0.701 total time=   2.7s
[CV 3/5; 1/9] START dropout_rate=0.0, learning_rate=0.001.......................
[CV 3/5; 1/9] END dropout_rate=0.0, learning_rate=0.001;, score=0.740 total time=   1.4s
[CV 4/5; 1/9] START dropout_rate=0.0, learning_rate=0.001.......................
[CV 4/5; 1/9] END dropout_rate=0.0, learning_rate=0.001;, score=0.745 total time=   2.5s
[CV 5/5; 1/9] START dropout_rate=0.0, learning_rate=0.001.......................
[CV 5/5; 1/9] END dropout_rate=0.0, learning_rate=0.001;, score=0.758 total time=   2.7s
[CV 1/5; 2/9] START dropout_rate=0.0, learning_rate=0.01........................
[CV 1/5; 

In [15]:
# Summarize the results
print("Best:{},using{}".format(grid_result.best_score_,grid_result.best_params_))
means=grid_result.cv_results_["mean_test_score"]
stds=grid_result.cv_results_["std_test_score"]
params=grid_result.cv_results_["params"]
for mean,stdev,param in zip(means,stds,params):
  print("{},{} with:{}".format(mean,stdev,param))

Best:0.7683049082756043,using{'dropout_rate': 0.2, 'learning_rate': 0.01}
0.7409133434295654,0.021163648470985695 with:{'dropout_rate': 0.0, 'learning_rate': 0.001}
0.765707504749298,0.039031608250831726 with:{'dropout_rate': 0.0, 'learning_rate': 0.01}
0.7448264241218567,0.023021645523628172 with:{'dropout_rate': 0.0, 'learning_rate': 0.1}
0.7384093165397644,0.042716286260454886 with:{'dropout_rate': 0.1, 'learning_rate': 0.001}
0.7669892191886902,0.02850185898372268 with:{'dropout_rate': 0.1, 'learning_rate': 0.01}
0.735684585571289,0.037036279129345394 with:{'dropout_rate': 0.1, 'learning_rate': 0.1}
0.6941261291503906,0.06690443093026233 with:{'dropout_rate': 0.2, 'learning_rate': 0.001}
0.7683049082756043,0.03585748965330314 with:{'dropout_rate': 0.2, 'learning_rate': 0.01}
0.7292674779891968,0.05912265207418094 with:{'dropout_rate': 0.2, 'learning_rate': 0.1}


- From the above results we can able to understand that from the Drop Out Rate (0.2) & Learning Rate (0.01) we got the best accuracy score is " Best:0.764417290687561 "

# **Hyperparameter Tuning for Activation Function & Kernel Intializer**

In [16]:
# Function to create and define the neural network model
def create_model(activation_function,init):
    model = Sequential()
    model.add(Dense(8, input_dim=8, kernel_initializer=init, activation=activation_function))
    model.add(Dropout(0.1))
    model.add(Dense(4,input_dim=8,kernel_initializer=init, activation=activation_function))
    model.add(Dropout(0.1))
    model.add(Dense(1, activation='sigmoid'))

    # Compiling the model
    adam = Adam(learning_rate=0.01)
    model.compile(loss='binary_crossentropy', optimizer=adam, metrics=['accuracy'])
    return model

# Create the model
model = KerasClassifier(build_fn=create_model, verbose=0, batch_size=40, epochs=10)

# Define the grid search parameters
activation_function= ["softmax","relu","tanh","linear"]
init = ["uniform","normal","zero"]

# Make a dictionary of grid search parameters
param_grids = dict(activation_function=activation_function,init=init)

# Build and fit the grid search CV
grid = GridSearchCV(estimator=model, param_grid=param_grids, cv=KFold(), verbose=10)
grid_result = grid.fit(X_standardized,y)

Fitting 5 folds for each of 12 candidates, totalling 60 fits
[CV 1/5; 1/12] START activation_function=softmax, init=uniform..................
[CV 1/5; 1/12] END activation_function=softmax, init=uniform;, score=0.740 total time=   1.9s
[CV 2/5; 1/12] START activation_function=softmax, init=uniform..................
[CV 2/5; 1/12] END activation_function=softmax, init=uniform;, score=0.714 total time=   1.8s
[CV 3/5; 1/12] START activation_function=softmax, init=uniform..................
[CV 3/5; 1/12] END activation_function=softmax, init=uniform;, score=0.773 total time=   2.1s
[CV 4/5; 1/12] START activation_function=softmax, init=uniform..................
[CV 4/5; 1/12] END activation_function=softmax, init=uniform;, score=0.843 total time=   2.8s
[CV 5/5; 1/12] START activation_function=softmax, init=uniform..................
[CV 5/5; 1/12] END activation_function=softmax, init=uniform;, score=0.765 total time=   2.0s
[CV 1/5; 2/12] START activation_function=softmax, init=normal...

In [17]:
# Summarize the results
print("Best:{},using{}".format(grid_result.best_score_,grid_result.best_params_))
means=grid_result.cv_results_["mean_test_score"]
stds=grid_result.cv_results_["std_test_score"]
params=grid_result.cv_results_["params"]
for mean,stdev,param in zip(means,stds,params):
  print("{},{} with:{}".format(mean,stdev,param))

Best:0.7683133959770203,using{'activation_function': 'softmax', 'init': 'normal'}
0.7670231819152832,0.043194395106629875 with:{'activation_function': 'softmax', 'init': 'uniform'}
0.7683133959770203,0.0392080007649872 with:{'activation_function': 'softmax', 'init': 'normal'}
0.747500216960907,0.06062782236836132 with:{'activation_function': 'softmax', 'init': 'zero'}
0.7605211853981018,0.04600156318886766 with:{'activation_function': 'relu', 'init': 'uniform'}
0.7579152822494507,0.05366505878160324 with:{'activation_function': 'relu', 'init': 'normal'}
0.6511586427688598,0.05244526932680711 with:{'activation_function': 'relu', 'init': 'zero'}
0.7657074928283691,0.038189463682198346 with:{'activation_function': 'tanh', 'init': 'uniform'}
0.7644172787666321,0.04787418723090888 with:{'activation_function': 'tanh', 'init': 'normal'}
0.6511586427688598,0.05244526932680711 with:{'activation_function': 'tanh', 'init': 'zero'}
0.7630761384963989,0.035951598287617265 with:{'activation_function

- From the above results we can able to see that the best accuracy score is given by activation_function -linear & init - uniform which is " Best:0.7761395454406739 "

# **Tuning of Hyperparameter - Number of Neurons in Activation Layer ❤**

In [18]:
# Function to create the neural network model
def create_model(neuron1, neuron2):
    model = Sequential()
    model.add(Dense(neuron1, input_dim=8, kernel_initializer="uniform", activation="tanh"))
    model.add(Dropout(0.2))
    model.add(Dense(neuron2, input_dim=neuron1, kernel_initializer="uniform", activation="tanh"))
    model.add(Dropout(0.1))
    model.add(Dense(1, activation='sigmoid'))

    # Compiling the model
    adam = Adam(learning_rate=0.001)
    model.compile(loss='binary_crossentropy', optimizer=adam, metrics=['accuracy'])
    return model

# Create the model
model = KerasClassifier(build_fn=create_model, verbose=0, batch_size=40, epochs=10)

# Define the Grid Search Parameters
neuron1 = [4, 8, 16]
neuron2 = [2, 4, 8]

# Make a dictionary of grid search parameters
param_grids = dict(neuron1=neuron1, neuron2=neuron2)

# Build and fit the grid search CV
grid = GridSearchCV(estimator=model, param_grid=param_grids, cv=KFold(), verbose=10)
grid_result = grid.fit(X_standardized, y)


Fitting 5 folds for each of 9 candidates, totalling 45 fits
[CV 1/5; 1/9] START neuron1=4, neuron2=2........................................
[CV 1/5; 1/9] END .........neuron1=4, neuron2=2;, score=0.753 total time=   2.3s
[CV 2/5; 1/9] START neuron1=4, neuron2=2........................................
[CV 2/5; 1/9] END .........neuron1=4, neuron2=2;, score=0.675 total time=   2.1s
[CV 3/5; 1/9] START neuron1=4, neuron2=2........................................
[CV 3/5; 1/9] END .........neuron1=4, neuron2=2;, score=0.740 total time=   1.5s
[CV 4/5; 1/9] START neuron1=4, neuron2=2........................................
[CV 4/5; 1/9] END .........neuron1=4, neuron2=2;, score=0.810 total time=   1.5s
[CV 5/5; 1/9] START neuron1=4, neuron2=2........................................
[CV 5/5; 1/9] END .........neuron1=4, neuron2=2;, score=0.752 total time=   1.5s
[CV 1/5; 2/9] START neuron1=4, neuron2=4........................................
[CV 1/5; 2/9] END .........neuron1=4, neuron2=4;,

In [19]:
# Summarize the results
print("Best : {}, using{}".format(grid_result.best_score_,grid_result.best_params_))
means=grid_result.cv_results_["mean_test_score"]
stds=grid_result.cv_results_["std_test_score"]
params=grid_result.cv_results_["params"]
for mean,stdev,param in zip(means,stds,params):
  print("{},{} with:{}".format(mean,stdev,param))

Best : 0.7669892191886902, using{'neuron1': 16, 'neuron2': 4}
0.7461845397949218,0.04305069325226764 with:{'neuron1': 4, 'neuron2': 2}
0.7435956239700318,0.04462640078432363 with:{'neuron1': 4, 'neuron2': 4}
0.7514302730560303,0.04833447388768534 with:{'neuron1': 4, 'neuron2': 8}
0.7501061081886291,0.0424949584691023 with:{'neuron1': 8, 'neuron2': 2}
0.7618368625640869,0.04647192633043428 with:{'neuron1': 8, 'neuron2': 4}
0.7565656542778015,0.03322401157994834 with:{'neuron1': 8, 'neuron2': 8}
0.7578898191452026,0.03324074359364308 with:{'neuron1': 16, 'neuron2': 2}
0.7669892191886902,0.038848171684146636 with:{'neuron1': 16, 'neuron2': 4}
0.7656990051269531,0.03425451496083485 with:{'neuron1': 16, 'neuron2': 8}


- From the above results we can able to see that the best accuracy score is given by neuron1 - 8 & neuron2 - 4 which is " Best : 0.7644172787666321 "

In [20]:
from sklearn.metrics import classification_report,accuracy_score

# Function to create the neural network model
def create_model():
    model = Sequential()
    model.add(Dense(16, input_dim=8, kernel_initializer="normal", activation="tanh"))
    model.add(Dropout(0.0))
    model.add(Dense(8, input_dim=16, kernel_initializer="normal", activation="tanh"))
    model.add(Dropout(0.0))
    model.add(Dense(1, activation='sigmoid'))

    # Compiling the model
    adam = Adam(learning_rate=0.01)
    model.compile(loss='binary_crossentropy', optimizer=adam, metrics=['accuracy'])
    return model

# Create the model
model = KerasClassifier(build_fn=create_model, verbose=0, batch_size=20, epochs=10)

#Fitting the model
model.fit(X_standardized,y)

#Predicting using trained model
y_predict=model.predict(X_standardized)

#Print the metrics
print(accuracy_score(y,y_predict))

24/24 [==============================] - 0s 1ms/step
0.7760416666666666


- With all the best values of Hyperparameters we got the overall accuracy score is 77%

Disadvantages
- High changes of face the overfitting problem in the situation of more Training accuracy & lesser Testing accuracy.

Best Ways to prevent the overfitting problem
- We can use the regularization techniques such as,
  Lasso (L1)
  Ridge (L2)
  ElasticNet

- We can also use the Dropout Techniques,by dropping the neurons from hidden layers and we can prevent the model for overfitting.

- We can use the Early Stopping to prevent the overfitting when the model is learn too much from the data.

- We have so many techniques by adding some noise to Training & Testing datasets.

- Batch Normalization

In [ ]:
def create_model(learning_rate,dropout_rate,activation_function,init,neuron1,neuron2):
    model = Sequential()
    model.add(Dense(neuron1,input_dim = 8,kernel_initializer = init,activation = activation_function))
    model.add(Dropout(dropout_rate))
    model.add(Dense(neuron2,input_dim = neuron1,kernel_initializer = init,activation = activation_function))
    model.add(Dropout(dropout_rate))
    model.add(Dense(1,activation = 'sigmoid'))

    adam = Adam(learning_rate = learning_rate)
    model.compile(loss = 'binary_crossentropy',optimizer = adam,metrics = ['accuracy'])
    return model

# Create the model

model = KerasClassifier(build_fn = create_model,verbose = 0)

# Define the grid search parameters

batch_size = [10,20,40]
epochs = [10,50,100]
learning_rate = [0.001,0.01,0.1]
dropout_rate = [0.0,0.1,0.2]
activation_function = ['softmax','relu','tanh','linear']
init = ['uniform','normal','zero']
neuron1 = [4,8,16]
neuron2 = [2,4,8]

# Make a dictionary of the grid search parameters

param_grids = dict(batch_size = batch_size,epochs = epochs,learning_rate = learning_rate,dropout_rate = dropout_rate,
                   activation_function = activation_function,init = init,neuron1 = neuron1,neuron2 = neuron2)

# Build and fit the GridSearchCV

grid = GridSearchCV(estimator = model,param_grid = param_grids,cv = KFold(),verbose = 10)
grid_result = grid.fit(X_standardized,y)

# Summarize the results
print('Best : {}, using {}'.format(grid_result.best_score_,grid_result.best_params_))
means = grid_result.cv_results_['mean_test_score']
stds = grid_result.cv_results_['std_test_score']
params = grid_result.cv_results_['params']
for mean, stdev, param in zip(means, stds, params):
  print('{},{} with: {}'.format(mean, stdev, param))

Fitting 5 folds for each of 8748 candidates, totalling 43740 fits
[CV 1/5; 1/8748] START activation_function=softmax, batch_size=10, dropout_rate=0.0, epochs=10, init=uniform, learning_rate=0.001, neuron1=4, neuron2=2
[CV 1/5; 1/8748] END activation_function=softmax, batch_size=10, dropout_rate=0.0, epochs=10, init=uniform, learning_rate=0.001, neuron1=4, neuron2=2;, score=0.649 total time=   2.4s
[CV 2/5; 1/8748] START activation_function=softmax, batch_size=10, dropout_rate=0.0, epochs=10, init=uniform, learning_rate=0.001, neuron1=4, neuron2=2
[CV 2/5; 1/8748] END activation_function=softmax, batch_size=10, dropout_rate=0.0, epochs=10, init=uniform, learning_rate=0.001, neuron1=4, neuron2=2;, score=0.584 total time=   2.5s
[CV 3/5; 1/8748] START activation_function=softmax, batch_size=10, dropout_rate=0.0, epochs=10, init=uniform, learning_rate=0.001, neuron1=4, neuron2=2
[CV 3/5; 1/8748] END activation_function=softmax, batch_size=10, dropout_rate=0.0, epochs=10, init=uniform, lear